File: backend/app/models/master_theme.py
from datetime import datetime

from sqlalchemy import JSON, DateTime, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.core.database import Base


class MasterTheme(Base):
    __tablename__ = "master_themes"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    concept: Mapped[str] = mapped_column(Text, nullable=False)

    character_json: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    environment_json: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    visual_style_json: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    animation_rules_json: Mapped[dict | None] = mapped_column(JSON, nullable=True)
    music_style_json: Mapped[dict | None] = mapped_column(JSON, nullable=True)

    reference_image: Mapped[str | None] = mapped_column(String(500), nullable=True)

    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, server_default=func.now(), onupdate=func.now()
    )

File: frontend/src/types/masterTheme.ts
export interface MasterTheme {
  id: number
  name: string
  concept: string
  character_json: Record<string, unknown> | null
  environment_json: Record<string, unknown> | null
  visual_style_json: Record<string, unknown> | null
  animation_rules_json: Record<string, unknown> | null
  music_style_json: Record<string, unknown> | null
  reference_image: string | null
  created_at: string
  updated_at: string
}

export interface MasterThemeInput {
  name: string
  concept: string
  character_json?: Record<string, unknown> | null
  environment_json?: Record<string, unknown> | null
  visual_style_json?: Record<string, unknown> | null
  animation_rules_json?: Record<string, unknown> | null
  music_style_json?: Record<string, unknown> | null
  reference_image?: string | null
}

export interface MasterThemeDraft {
  name: string
  concept: string
  character: Record<string, unknown>
  environment: Record<string, unknown>
  visual_style: Record<string, unknown>
  animation_rules: Record<string, unknown>
  music_style: Record<string, unknown>
}
